### ISSUE TITLE
[Bug]: Unintuitive matching of line styles and labels in legend

### ISSUE THREAD

**Issue by maxaehle** [matplotlib/matplotlib#27725]
### Bug summary

In a figure with one `plt.errorbar` and one `plt.plot` line, legend text must be supplied in a counterintuitive order.

### Code for reproduction

```python
import matplotlib.pyplot as plt
plt.errorbar([0,1],[0,1],[.3,.3])
plt.plot([0,1],[1,0])
plt.legend(["increasing","decreasing"])
plt.show()
```

### Actual outcome

The legend associates the `plt.plot` line with `increasing` and the `plt.errorbar` line with the label `decreasing`.

![actual_outcome](https://github.com/matplotlib/matplotlib/assets/72806890/b2164465-01fc-4bd2-adac-4f78f989c7e7)

### Expected outcome

The two line styles in the legend should be swapped. I.e., the `plt.errorbar` line should be associated with the label `increasing` and the `plt.plot` line with the label `decreasing`.

### Additional information

The user can work around the issue by providing the legend text to `plt.errorbar`/`plt.plot` via the keyword argument `label`.

The [documentation](https://matplotlib.org/3.8.2/api/_as_gen/matplotlib.axes.Axes.legend.html#matplotlib.axes.Axes.legend) of `legend` says about the way we use it (supplying only label text) that

> This call signature is discouraged, because the relation between plot elements and labels is only implicit by their order and can easily be mixed up.

### Operating system

Ubuntu 22.04

### Matplotlib Version

3.5.1

### Matplotlib Backend

TkAgg

### Python version

3.10.12

### Jupyter version

N/A

### Installation

None

**Comment by rcomer** [comment]
I think the example you give demonstrates why that approach is discouraged in the documentation.  If you do not want to pass the labels to the plotting methods, another option would be to pass the outputs as legend handles:

```python
import matplotlib.pyplot as plt
err = plt.errorbar([0,1],[0,1],[.3,.3])
line, = plt.plot([0,1],[1,0])
plt.legend([err, line], ["increasing","decreasing"])
plt.show(
```
![image](https://github.com/matplotlib/matplotlib/assets/10599679/896d14d1-1815-490b-a39b-71ac435f8c48)

**Comment by rcomer** [comment]
Perhaps I missed the point above.  Where the docs say "the relation between plot elements and labels is only implicit by their order", that could be taken to mean that they are defined by the order they were added.  That is clearly not the case here.

The relevant code is
https://github.com/matplotlib/matplotlib/blob/41d4149f84b40d3a90a70f81b940985b8485f04a/lib/matplotlib/legend.py#L1251-L1254

`errorbar` returns a container and `plot` returns a `Line2D` artist.  This loop always puts containers last.  So maybe the documentation should be reworded for clarity?

**Comment by anntzer** [comment]
We could also detect when both sublists are non-empty and emit a warning in that case (suggesting the more explicit ordering method), as the order is likely to be messed up otherwise.

**Comment by timhoffm** [comment]
A warning at runtime is certainly helpful. I’m tempted to deprecate labels-only completely. The documentation will get harder to understand with correct description of container order. And even without the container order issue, this is a potential footgun.

**Comment by mwaskom** [comment]
> I’m tempted to deprecate labels-only completely. 

Please do this! People get themselves into so much trouble trying to customize seaborn legends this way, and I don't really know what to tell them.

**Comment by mwaskom** [comment]
The official seaborn line is that you should pass in the values that you want to appear in the legend. That’s not ideal and I would prefer it if the matplotlib legend were modifiable, but in a more explicit way (eg something like `legend.set_labels()`).  But even with that sort of API, I think it’s risky to assume that the order of artists is going to be what you expect; if seaborn doesn’t use the order you expect, your plot will be wrong in a non-obvious way. But it’s still better than the case where you *think* `ax.legend(labels)` is just replacing the labels, but it’s actually drawing a legend from scratch by grabbing the first artists attached to the axes.

**Comment by jklymak** [comment]
Can people not do `artist.set_label`?

**Comment by mwaskom** [comment]
Yeah I guess, but for seaborn there's not really much benefit to that over calling `rename` on the dataframe you're sending into the plot. For matplotlib usecases I think it's probably fine to have `.set_labels(labels: list[str])`.

**Comment by rcomer** [comment]
> > I’m tempted to deprecate labels-only completely.
> 
> It's really helpful if you're using something like pandas and want to customize the labels, where you really might not want to use those labels for your columns.

If you know what the auto-labels will be but do not like them, can you do something like

```python
my_map = {'old1':'new1', 'old2':'new2'}

handles, labels = ax.get_legend_handles_labels()
ax.legend(handles, [my_map[l] for l in labels])
```
?

**Comment by jklymak** [comment]
That trick above is OK, but pretty undiscoverable.  I think `legend.remap/rename_labels` per https://github.com/matplotlib/matplotlib/issues/27725#issuecomment-1920452030  is relatively simple and not likely to lead to confusion.

**Comment by timhoffm** [comment]
> That trick above is OK, but pretty undiscoverable. I think `legend.remap/rename_labels` per [#27725 (comment)](https://github.com/matplotlib/matplotlib/issues/27725#issuecomment-1920452030) is relatively simple and not likely to lead to confusion.

May be ok. We need to make clear though that legend labels are not linked to the artist labels. They are extracted once on legend creation. So renaming a legend label does not affect the Artist label.

If you want to implement this, I suggest to follow the [pandas rename](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html) and accept a dict or a function. Also for convenience, we should accept both a label (str) or a handle (Artist) as key.
